hardware/raspberrypi/bootmodes/pxetools/pxetools

```diff
@@ -0,0 +1,229 @@
+#!/usr/bin/env python3
+
+import os
+import re
+import subprocess
+import time
+import sys
+from tabulate import tabulate
+
+HELP = "Usage:\n\
+\tpxetools --add serial\n\
+\tpxetools --remove serial\n\
+\tpxetools --list\n"
+
+LAN = "10.3.14.0/24"
+NFS_IP = "10.3.14.18"
+
+def cmd(cmd, print_out=False, print_cmd=True, can_fail=False):
+    if print_cmd:
+        print("# {}".format(cmd))
+
+    out = None
+    
+    try:    
+        out = subprocess.check_output(cmd, shell=True).decode()
+    except:
+        if not can_fail:
+            raise
+
+    if print_out:
+        print(out)
+
+    return out
+        
+
+def add():
+    serial = None
+    if len(sys.argv) != 3:
+        if os.path.exists("/usr/local/sbin/get_serial"):
+            # Auto detect
+            print("Will try and auto detect serial. Please plug in your Pi now!")
+            cmd("systemctl stop dnsmasq")
+            cmd("iptables -t raw --flush")
+
+            try:
+                serial = cmd("get_serial", print_cmd=False).rstrip()
+            except KeyboardInterrupt:
+                cmd("iptables-restore < /etc/iptables/rules.v4")
+                cmd("systemctl start dnsmasq")
+                sys.exit()
+
+            print("Found serial")
+            cmd("iptables-restore < /etc/iptables/rules.v4")
+            cmd("systemctl start dnsmasq")
+    else:
+        serial = sys.argv[2]
+    
+    print("Serial: {}".format(serial))
+    owner = input("Owner Name (ex Gordon): ")
+    name = input("Name for pi: ")
+    
+    print("Select a base image:")
+    selection = ["I will prepare my own filesystem"] + os.listdir('/nfs/bases')
+    for i in range (0, len(selection)):
+        print("\t{}. {}".format(i + 1, selection[i]))
+    
+    img_choice = input("Enter an option number: ")
+
+    # Validate stuff
```

Comment:
All this validation of the `serial` value should probably be done **before** asking the user for input?